servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/DefaultAsyncContextProvider.java

```diff
@@ -240,82 +240,106 @@ public <T> Subscriber<T> wrapPublisherSubscriber(final Subscriber<T> subscriber,
         return ContextPreservingScheduledExecutorService.of(executor);
     }
 
     @Override
-    public <T> CompletableFuture<T> wrapCompletableFuture(final CompletableFuture<T> future, final ContextMap context) {
+    public final <T> CompletableFuture<T> wrapCompletableFuture(final CompletableFuture<T> future,
+                                                                final CapturedContext context) {
         return ContextPreservingCompletableFuture.newContextPreservingFuture(future, context);
     }
 
     @Override
-    public Runnable wrapRunnable(final Runnable runnable, final ContextMap context) {
+    public final Runnable wrapRunnable(final Runnable runnable, final CapturedContext context) {
         return new ContextPreservingRunnable(runnable, context);
     }
 
     @Override
-    public <V> Callable<V> wrapCallable(final Callable<V> callable, final ContextMap context) {
+    public final <V> Callable<V> wrapCallable(final Callable<V> callable, final CapturedContext context) {
         return new ContextPreservingCallable<>(callable, context);
     }
 
     @Override
-    public <T> Consumer<T> wrapConsumer(final Consumer<T> consumer, final ContextMap context) {
+    public final <T> Consumer<T> wrapConsumer(final Consumer<T> consumer, final CapturedContext context) {
         return new ContextPreservingConsumer<>(consumer, context);
     }
 
     @Override
-    public <T, U> Function<T, U> wrapFunction(final Function<T, U> func, final ContextMap context) {
+    public final <T, U> Function<T, U> wrapFunction(final Function<T, U> func, final CapturedContext context) {
         return new ContextPreservingFunction<>(func, context);
     }
 
     @Override
-    public <T, U> BiConsumer<T, U> wrapBiConsumer(final BiConsumer<T, U> consumer, final ContextMap context) {
+    public final <T, U> BiConsumer<T, U> wrapBiConsumer(
+            final BiConsumer<T, U> consumer, final CapturedContext context) {
         return new ContextPreservingBiConsumer<>(consumer, context);
     }
 
     @Override
-    public <T, U, V> BiFunction<T, U, V> wrapBiFunction(final BiFunction<T, U, V> func, final ContextMap context) {
+    public final <T, U, V> BiFunction<T, U, V> wrapBiFunction(
+            final BiFunction<T, U, V> func, final CapturedContext context) {
         return new ContextPreservingBiFunction<>(func, context);
     }
 
+    private static final class CapturedContextImpl implements CapturedContext {
+
+        private final ContextMap contextMap;
+
+        CapturedContextImpl(ContextMap contextMap) {
+            this.contextMap = contextMap;
+        }
+
+        @Override
+        public ContextMap captured() {
+            return contextMap;
+        }
+
+        @Override
+        public Scope restoreContext() {
+            ContextMap prev = exchangeContext(contextMap);
+            return NO_DEBUG_LOGGING && prev instanceof Scope ? (Scope) prev : () -> detachContextMap(contextMap, prev);
```

Comment:
Could you please use an inner static class instead of a lambda here? It's easier to read flame graphs and heap dump with well-named classes instead of lambdas with a counter.